### ISSUE TITLE
[HardwareSim] Configuration to enable/disable LCM (to support three use cases)

### ISSUE THREAD

**Issue by RussTedrake** [RobotLocomotion/drake#20055]
I would like to try replacing my manipulation station workflows using the hardware sim schemas / loaders.  Ideally I would like to parse the same yaml file three times:
1) stand-alone simulation. This is what `hardware_sim.cc` / `hardware_sim.py` is doing now.  That's great!  (It replaces my existing `examples/manipulation_station/mock_station_simulation.cc`.  
2) to create the equivalent of `ManipulationStation` for in-process simulation.  I don't want any LCM here; instead I want to Export the driver input and output ports so that the diagram can be used as a component in another diagram.
3) to create the equivalent of `ManipulationStationHardwareInterface`.  Rather than load the plant and the drivers directly, I want to create a diagram which implements the complement. E.g. `hardware_sim` subscribes to `IIWA_COMMAND`, in this workflow I want the interface to be a diagram that _publishes_ the `IIWA_COMMAND` directly from an input port.

Use case (2) should be able to reuse most of the same `ApplyDriverConfig` methods that we use in (1), but currently `ApplyDriverConfig` always adds the lcm publishers and subscribers.  I could go through and them, but it would be better to be able to opt-out in my calls to ApplyDriverConfig, and (ideally) export the ports instead.

Use case (3) probably involves a different, but closely coupled `Apply_?_Config` code path.  It would be nice to consider this in the design as well.

**Comment by RussTedrake** [comment]
I've worked around this locally -- you can see the draft [here](https://github.com/RussTedrake/manipulation/blob/master/manipulation/station.py).  But I do think that this workflow is something that is worth supporting upstream, and my version is not yet able to use e.g. the logic in Drake for making the controller_plant.

**Comment by jwnimmer-tri** [comment]
Item (3) is basically RobotHub from Anzu.  See `anzu/robot_bridge/hub/hub_builder_config.h` and nearby.

**Comment by RussTedrake** [comment]
I took a look, and I agree.  Although I find it more elegant to have the same config file be able to support all three cases?

I've got a narrow version working at the link I shared above, but would love to promote the HardwareInterface / HubBuilder idea to Drake.

**Comment by jwnimmer-tri** [comment]
For (2), I think the best path will be to refactor `build_iiwa_control.cc` to add the control cabinet simulation as a separate System (currently named `SimIiwaDriver` as a placeholder), and connect the LCM I/O to those system ports.  Then, we can offer a flag (or something) to not connect the LCM.  The nice cabinet sim System will remain, which you can export the ports from.

The WIP branch is [here](https://github.com/jwnimmer-tri/drake/commits/iiwa-builder-subdiagram) with [SimIiwaDriver here](https://github.com/jwnimmer-tri/drake/blob/iiwa-builder-subdiagram/manipulation/kuka_iiwa/sim_iiwa_driver.h).

Here's a picture of `hardware_sim` running with the new refactoring (with LCM still enabled).

![image](https://github.com/RobotLocomotion/drake/assets/17596505/3d534bbb-7a65-4113-96be-c8b0a1acd983)

**Comment by RussTedrake** [comment]
This is great!  Thank you.  

See the ManipulationStation ports [here](https://drake.mit.edu/doxygen_cxx/classdrake_1_1examples_1_1manipulation__station_1_1_manipulation_station.html).  A few lessons from that work-flow:
- I think we should prefix the model instance name on the SimIiwaDriver ports.
- I think it's valuable to add the "cheat" ports, at least when not in LCM mode.  Probably adding them always is fine, since they won't be evaluated.

It's not obvious to me that this refactor is needed to make the lcm flag work, but it seems very reasonable to me.

**Comment by jwnimmer-tri** [comment]
> I think we should prefix the model instance name on the SimIiwaDriver ports.

This could happen when the ports are exported out of the outer builder.  Within any one `SimIiwaDriver` system, an unprefixed port name is sufficient.  The `SimIiwaDriver` itself is named per the model instance.  This is consistent with e.g. `IiwaStatusReceiver` whose port names are just `position_measured` not `{model_name}_position_measured`.

For reference, in robot hub, we use `{model_name}.{port_name}` when exporting, i.e., a dot instead of underscore.  I think it improves clarity (and makes it easier to mechanically parse) when the model name vs port name use a different separator (.) than the separator for words within a port name (_).

When I next post a refreshed image, I'll use a left arm and vs right arm as a better example.

> I think it's valuable to add the "cheat" ports, at least when not in LCM mode. Probably adding them always is fine, since they won't be evaluated.

The (C++) manipulation station doesn't have any per-model-instance cheat ports, does it?  Any per-plant cheat ports would be exported out of the plant, not out of the per-model-instance sim driver.

**Comment by RussTedrake** [comment]
Makes sense.

> The (C++) manipulation station doesn't have any per-model-instance cheat ports, does it? Any per-plant cheat ports would be exported out of the plant, not out of the per-model-instance sim driver.

That's correct.  I hadn't thought through the implications of the sim driver being per-model-instance.

I'm fine with `{model_name}.{port_name}`.  It would be good to rip that bandaid off before the fall terms starts, if possible to do it this week. Perhaps I should make that change in my hardware station draft right now?

**Comment by jwnimmer-tri** [comment]
> It's not obvious to me that this refactor is needed to make the lcm flag work, but it seems very reasonable to me.

The subdiagram allows us to give semantic names to intermediate signals, so we can have "torque commanded" instead of "adder output".  That should provide the lever we'll need to mechanically export the cabinet sim ports in use case (2), without relying on species-specific logic.

> Perhaps I should make that change in my hardware station draft right now?

Probably a good idea, assuming you buy the premise.

The other thing we worked on some with robot hub was to make sure the port names were consistent across species.  I suggest keeping an eye out for any unwanted variance there, as well.

**Comment by RussTedrake** [comment]
Re: naming intermediate signals.  Makes sense.  Thanks.

Re: "_" -> "."  Sounds good.  Will make the change in my course repo now.

The more I've thought about it, the more I feel certain that I'd like to use the same scenario file for all three workflows.  Do you agree on that premise?

**Comment by jwnimmer-tri** [comment]
> Do you agree on that premise?

It sounds appealing, but I'm not sure whether it'll work.  The attributes of a robot driver will vary a lot (with little if any overlap) between the three use cases.   For a directly simulated robot (case 2), there will be knobs to configure how the commands turn into dynamics (e.g., gains).  For commanding hardware (case 3), there will be knobs to configure message rates and timeout / failsafe limits (like we see in the current robot hub driver config structs).  For emulating hardware (case 1), it's a blend of the sim config (case 2) plus the _complementary_ side of the message timing.

One way to attack that, while keeping the Scenario as just one uniform dataclass, would be to have the driver `Union[]` have options for both IiwaDriver vs IiwaHardwareDriver, so the yaml edit to switch from sim to hardware would be to edit each model's stanza.  I'm not sure whether there's a way to just have `Union[IiwaDriver, WsgDriver]` and somehow choose hardware vs sim elsewhere (e.g. an argparse flag) while applying the scenario.

I could possibly see case 1 and case 2 sharing almost everything, though.  I've been thinking of either allowing `LcmBuses` to be nullptr to indicate "no LCM", or else switch over to `std::optional<std::string> lcm_bus{"default"};` in the driver config (i.e., add `optional<>` so the user could say `lcm_bus: null` in the yaml to turn off LCM one driver at a time.

**Comment by RussTedrake** [comment]
The argument for Case 2 and Case 3 sharing a file is the "flip one switch to go from in-process simulation to controlling the robot".  That's an important property if we can obtain it.  Adding additional attributes to the yaml file feels reasonable.  Parsing an entirely separate file feels like it corrupts the intended abstraction.

**Comment by jwnimmer-tri** [comment]
Please check the following figure, to see if it meets the requirement for cases (1) and (2).

Note that in this figure I've split the MbP system in half for clarity (outputs vs inputs).

For case (1) the `ApplyDriverConfig(IiwaDriver)` would create all of the systems in this figure (except for the MbP+SG).  This is the same as what we have today, except for the introduction of the inner Diagram wrapper around some of the systems.

For case (2) the `ApplyDriverConfig(IiwaDriver)` would omit everything marked in red.  In other words, it would only add the `SimIiwaDriver` sub-diagram, and connect it to the plant (the bold arrows).

Note in particular that it would _not_ export any ports on the main diagram.  That would be up to the user's application to decide whether they should be exported or connected.  However, by using the nested diagram for the sim driver, all of the ports already have useful names, so it should be easy to (mechanically) find the driver system and blanket connect and/or export all of the ports generically, without any species-specific control flow.

---

![image](https://github.com/RobotLocomotion/drake/assets/17596505/4bc923af-b2b3-4936-a7bb-e72ac2118857)

**Comment by pangtao22** [comment]
How can we give feedback / feature request to the new HardwareSimulation ? For example, right now the iiwa controller added in `ApplyDriverConfigSim` is the [inverse dynamics controller with hard-coded gains](https://github.com/RussTedrake/manipulation/blob/07fb9bff42d4a0c09c5bf6b0f3dd5f54026b533e/manipulation/station.py#L195-L203), but we know the iiwa actually uses a [joint stiffness controller](https://github.com/RobotLocomotion/drake/issues/11955). In fact, with the recent merge of implicit PD controller into main, we should be given the option to use implicit PD + gravity compensation as well. 

More generally, it would be nice to be able to specify in `FooDriver` user-defined controller and its parameters. This can be useful on robot platforms for which we have joint-level access, as we typically will write some low-level controller ourselves.

**Comment by jwnimmer-tri** [comment]
> Please check the following figure, to see if it meets the requirement for cases (1) and (2).

@RussTedrake still waiting for your feedback about this design.  The first PR is underway in #20182.

> How can we give feedback / feature request to the new HardwareSimulation?

Opening feature request issues is fine.  We have some already (#20165, #20052, #20061, #20058, #20054, #20053).  If you're not sure whether a feature might already exist, you're also welcome to discuss on slack.

> ... right now the iiwa controller ... is the IDC but we know the iiwa actually uses a JSC...

What I have in mind for the `IiwaDriver` in `drake/manipulation/kuka_iiwa` is to have the *best* simulation model available, with config settings for fine tuning.  For example we already have `ext_joint_filter_tau` config setting and there is a `TODO(jwnimmer-tri) Make desired_iiwa_kp_gains configurable.` to add more settings.  If we think a different controller is better than IDC, then we should switch our `IiwaDriver` implementation to use it.  If there are inescapably multiple kinds of controllers used under different circumstances to act like an iiwa, then we could imagine offering a setting in `IiwaDriver` for which one to use, but I would really like to avoid that.  Since users can write their own drivers, it's not important that we offer every single option.  My aim is to offer one, best, nominal driver option.

> More generally, it would be nice to be able to specify in FooDriver user-defined controller and its parameters. This can be useful on robot platforms for which we have joint-level access, as we typically will write some low-level controller ourselves.

This is already possible, if I understand you correctly.  You can already write any driver you want, and name it anything you want (even `IiwaDriver`), with any control policy you want.  However, we don't have good examples that show how to do this yet.  See #20052.

**Comment by pangtao22** [comment]
I'm pretty sure that the InverseDynamics controller we're using right now is not what iiwa uses as its internal controller. According to [Ott2008](https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=4451347), the closed-loop dynamics of the iiwa is:
<img width="595" alt="image" src="https://github.com/RobotLocomotion/drake/assets/20274938/c4e5d297-be8e-4771-b681-2a69743f4d2e">
which is different from the inverse dynamics controller. The difference is negligible when the robot moves in free-space, but is critical when robots deliberately makes contact to exert contact forces. 

Russ's JointStiffnessController was written to address this issue, but it suffers from small simulation steps when the joint stiffness is set to realistic values. I think the new implicit PD controller is the right way to model iiwa's internal controller.

**Comment by RussTedrake** [comment]
@pangtao22 -- i think the stiffness control is definitely a good discussion (and you know I agree with your premise), but not immediately relevant to this issue.

> Please check the following figure, to see if it meets the requirement for cases (1) and (2).

BTW - the sign flip is news to me.  where did that come from??

The main question I have is about not exporting the input/output ports.  My goal (which I'm happy to accomplish partly in user code), is to have ports like `iiwa.position_measured` on the resulting diagram.  But it could instead be, `wsg.position`. In particular, the ports to be exported _would_ depend on the species.  Could you help me understand what you meant by not needing any species-specific glue code?

**Comment by jwnimmer-tri** [comment]
> BTW - the sign flip is news to me. where did that come from??

It's been there forever.  Here's the spelling on master:

https://github.com/RobotLocomotion/drake/blob/6fec0178a2cd7b026d6a05b1ef377b1543af283c/manipulation/kuka_iiwa/build_iiwa_control.cc#L235-L238

I kept it on the LCM side of the partition, assuming it's wrong and we're going to remove it at some point.  (The C++ manipulation_station does not have it.)

> The main question I have is about not exporting the input/output ports. My goal (which I'm happy to accomplish partly in user code), is to have ports like iiwa.position_measured on the resulting diagram. But it could instead be, wsg.position. In particular, the ports to be exported would depend on the species. Could you help me understand what you meant by not needing any species-specific glue code?

Something like this in your scenario loader:

```py
# Add simulated drivers (with no LCM).
ApplyDriverConfigs(
    driver_configs=scenario.model_drivers,
    sim_plant=sim_plant,
    models_from_directives=added_models,
    lcm_buses=None,
    builder=builder)
for model_name, driver_config in scenario.model_drivers.items():
    # Find the subsystem with the driver for this model instance.
    driver_type = type(driver_config).name
    driver_subsystem = builder.GetSubsystemByName(f"{driver_type}({model_name})")
    # Export any unconnected input ports.
    for i in range(driver_subsystem.num_input_ports()):
        port = driver_subsystem.get_input_port(i)
        if not builder.IsConnectedOrExported(port):
            builder.ExportInput(port, f"{model_name}.{port.get_name()}")
    # Export any unconnected output ports.
    ... likewise ...
```

**Comment by RussTedrake** [comment]
Oh.  That's clever.  I didn't know about `IsConnectedOrExported`.  In that case, the design LGTM.  Thank you.

**Comment by jwnimmer-tri** [comment]
The `SimIiwaDriver` refactoring has landed.  In a pinch, that's enough to make a non-LCM diagram.  (Call the `ApplyDriverConfig` which adds everything, and then call `builder.RemoveSystem` on the all but the first newly-added system so that only the subdiagram remains.)

Coming next up will be to offer the flag to not even add LCM in the first place.

**Comment by RussTedrake** [comment]
Thanks!

**Comment by jwnimmer-tri** [comment]
@RussTedrake one more question.

When you say "disable LCM" is that only for the robot controllers or is it for _everything_ -- including visualization messages.  The architecture will support the end-user choosing either way, but your goal will probably affect how I round out the examples.

**Comment by RussTedrake** [comment]
I would have said "for everything", but that's because I don't use DrakeVisualizer/meldis.  I agree that visualization messages could be the exception.  The goal for the route 1 workflow is to do whatever we would do if we were setting up the simulation all in a single process.

**Comment by RussTedrake** [comment]
The refactoring done to address this issue so far makes it _possible_ to opt out of LCM and write the in-process sim.  Now I would like to make sure we agree on how to make the in-process sim a first class citizen.

In my manipulation repo, the pattern I've been converging towards is to make [thin wrappers](https://github.com/RussTedrake/manipulation/blob/f84ec8f4769866d4b4b9ee1f3ab4158323074346/manipulation/station.py#L919) around the existing Apply* methods, e.g.
```
def _ApplyCameraConfigSim(
    *, config: CameraConfig, builder: DiagramBuilder, lcm_buses: LcmBuses
) -> None:
    # Always opt-out of lcm in this workflow.
    this_config = copy(config)
    this_config.lcm_bus = "opt_out"
    ApplyCameraConfig(config=this_config, builder=builder, lcm_buses=lcm_buses)

    camera_sys = builder.GetSubsystemByName(f"rgbd_sensor_{config.name}")
    builder.ExportOutput(
        camera_sys.color_image_output_port(), f"{config.name}.rgb_image"
    )
    builder.ExportOutput(
        camera_sys.depth_image_32F_output_port(), f"{config.name}.depth_image"
    )
    builder.ExportOutput(
        camera_sys.label_image_output_port(), f"{config.name}.label_image"
    )
```

I could see a few paths forward
1) two methods in Drake (e.g. AddCameraConfig becomes a variant for in-process vs remote sim)
2) pass a new argument into the existing AddCameraConfig to switch between in-process vs remote.

I dislike having the name AddCameraConfig assume the remote sim case (in my view the in-process sim should be at least as dominant of a use case, perhaps more dominant in Drake).

I'd like to move forward on this asap.  Any thoughts @jwnimmer-tri , @IanTheEngineer , @ggould-tri , @dmcconachie-tri ?

**Comment by jwnimmer-tri** [comment]
Sending LCM isn't necessarily about "remote sim" -- it's extremely useful to have LCM messages sent while debugging an in-process sim, for example.  Another common use is to use the LCM messages as the log format for creating simulation traces.

For users that never want LCM, what I imaged they would do is in their `class Scenario`, set the `default` LcmBus to null.  Something like this:

```patch
--- a/examples/hardware_sim/hardware_sim.py
+++ b/examples/hardware_sim/hardware_sim.py
@@ -88,7 +88,8 @@ class Scenario:
     # A map of {bus_name: lcm_params} for LCM transceivers to be used by
     # drivers, sensors, etc.
     lcm_buses: typing.Mapping[str, DrakeLcmParams] = dc.field(
-        default_factory=lambda: dict(default=DrakeLcmParams()))
+        default_factory=lambda: dict(default=DrakeLcmParams(
+            lcm_url=LcmBuses.kLcmUrlMemqNull)))
 
     # For actuated models, specifies where each model's actuation inputs come
     # from, keyed on the ModelInstance name.
```

In other words, the LCM modeling choice is a feature of the _scenario_, not an out-of-band knob.  If the user does or doesn't want LCM, their scenario should say so.

**Comment by RussTedrake** [comment]
That argument might hold for output ports, but it does not hold for input ports.  If I connect an LcmSubscriber to an input port, then I cannot connect to it directly in my diagram.

**Comment by jwnimmer-tri** [comment]
Your example wasn't with input ports, so I'm not clear what you problem you are trying to solve.

As I understand it, it should be possible:

(1) to set defaults to your liking in your own `class Scenario` (so that each individual `my_scenario.yaml` doesn't need to repeat them), and

(2) when necessary, to programmatically munge `scenario` object to en/disable specific LCM behavior (e.g., opt-in to LCM visualization) as part of your outermost setup code (i.e., at the level of `main()`),

... and with only those two tools, you can meet all the requirements without touching or wrapping any of the Apply functions, and without any gross code.

Do you have an example where those two tools are not sufficiently capable, or not sufficiently clean?

**Comment by RussTedrake** [comment]
> Your example wasn't with input ports, so I'm not clear what you problem you are trying to solve.

Consider the case of the IiwaDriver for a more holistic view of the problem.

In the single-process diagram workflow, it would be wrong to ever connect an LcmSubscriber to the controller input ports. The user should not be able to cause that wiring to happen from any version of their yaml file.

As I said:
> The refactoring done to address this issue so far makes it possible to opt out of LCM and write the in-process sim. Now I would like to make sure we agree on how to make the in-process sim a first class citizen.

There are three use cases for loading from yaml. Currently, the Apply* functions assume the hardware_sim workflow, with lcm, as the default and people have to actively opt out (and the logic is non-trivial; i got it wrong at least 3 times in the last 24 hours; for instance Anzu's ApplyCameraConfig python binding only takes a LcmInterfaceSystem, and Drake's takes DrakeLcmInterface).  Your proposal is that every downstream consumer would have to re-implement this opt-ing out and the ExportInput / ExportOutput methods required to make a consistent interface with the iiwa.  You argue that lcm can always be helpful.

In my mind, this is backwards.  I feel that ExportInput/ExportOutput should always be done; and the LCM (if wanted) could be opted-in, and attached to the exported ports.  But that's a matter of opinion.  Your view makes the remote development the primary use case. I believe single process is more fundamental.

My goal is to have one set of code which exports the correct, consistently named input and output ports for the iiwa, and have that live in Drake.  Of course I want this for all of the Apply* methods, not just the iiwa.  Already i am copy pasta-ing lots of code from manipulation to anzu, and working through the slightly different api touchpoints.  I don't want to maintain even two copies of every feature.

**Comment by jwnimmer-tri** [comment]
My view is that everything that ends up in the diagram should be _denoted in the Scenario object_.  There must be no "mode" side-band flag that changes how the diagram ends up.  A scenario means exactly one thing, i.e., a specific diagram.  Therefore the "don't connect the control inputs to LCM" is not a different family of apply functions, but rather a different `scenario` instance compared to when LCM is connect to the control inputs.

If we need to _programatically_ convert a no-control-input scenario instance to an LCM-controlled scenario instance, we do that by mutating the scenario object and then calling the same Apply functions as usual, not by having different family or Apply functions or by adding new arguments to the Apply functions beyond just their Config instance.

>  Your view makes the remote development the primary use case. I believe single process is more fundamental.

Absolutely nowhere am I advocating for the controller to live in a separate process by default!

I would, however, advocate that "log the simulation to LCM for after-the-fact inspection" is an important flow and should be very easy.

> Currently, the Apply* functions assume the hardware_sim workflow, with lcm, as the default and people have to actively opt out ...

No, they don't.

The examples/hardware_sim has LCM control enabled by default.

If you want _your_ scenario's semantics to be opt-in to LCM instead of opt-out to LCM, then define your scenario to default to the null LCM bus.  (I illustrated that just above.)  The Apply functions have  _no opinion_ about whether LCM is on our off by default.  They just do whatever their `Config` object tells them to do.  Their config struct's default LCM is named "default", so there is a single point of control in the LCM buses for what LCM should be doing.

> Anzu's ApplyCameraConfig python binding only takes a LcmInterfaceSystem, and Drake's takes DrakeLcmInterface

Inevitably, some code we have today is broken / not up to par.  We should fix that code to match our current habits.

> Your proposal is that every downstream consumer would have to re-implement this opt-ing out and the ExportInput / ExportOutput methods required to make a consistent interface with the iiwa.

Keep a clear separation in your mind about _driver_ systems versus other, publish-only systems.

In the case of publish-only systems, acute handling of the "null bus" is roughly speaking just an optimization.  If they are null-agnostic and still publish messages to the null bus, those messages will just be silently dropped without complaint.  (The downside would be possibly-unwanted publish events, and the associated compute cost.)

For drivers, if the driver knows how to speak LCM _sometimes_, then it needs to have some config-struct mechanism to turn that on or off.  Instead of adding a specific new "on/off" field onto each driver, instead I chose to reuse the "lcm_bus" name for this purpose.  The config says which "semantic" bus the driver should be using, and that resolves to either null or real.  That allows for a single point of control to turn on/off _all_ drivers at once.

In other words, I don't understand your complaint here.  If I driver knows about LCM at all, then it needs to know about how to turn it on/off.  There will _necessarily_ need to be _some_ kind of code in each driver that handles that choice.  I don't think it's a defect in the design here that each driver needs a non-zero amount of code to handle the case when LCM is disabled.

Also keep in mind: the Apply functions need not be responsible for exporting ports.  We can export the ports all at once in bulk, after all of the Apply functions have finished, as part of the `main`-level function.  (The recipe was shown upthread -- programatically export all outputs, and all disconnected inputs).

**Comment by RussTedrake** [comment]
> My view is that everything that ends up in the diagram should be denoted in the Scenario object. There must be no "mode" side-band flag that changes how the diagram ends up. A scenario means exactly one thing, i.e., a specific diagram. Therefore the "don't connect the control inputs to LCM" is not a different family of apply functions, but rather a different scenario instance compared to when LCM is connect to the control inputs.

That's helpful, but I'm not quite onboard with that viewpoint yet. I think that a single Scenario yaml file should be sufficient for launching all three use cases.  Why require people to duplicate their scenario three times (and possibly introduce errors in the process)?  I agree that we could support even more than 3 use cases in the long run with that sort of design. But these three are the important ones, and have been the important ones since the early days of the manipulation station example.  Pro grammatically modifying the yaml into three variants on disk before loading seems clunky and error prone to me.

My code also exports all outputs and all unconnected input ports. But currently I am using e.g. the ExportInput to rename a generically named input port on a system into the known message types associated with the IiwaDriver (e.g. https://manipulation.mit.edu/robot.html#section5) .  Where would you introduce that renaming, if not at the `ExportInput` level?  Of course one could add extra systems to do it, but that again seems awfully clunky (in the setup, but also in the resulting graphviz, etc)?

**Comment by jwnimmer-tri** [comment]
> Why require people to duplicate their scenario three times (and possibly introduce errors in the process)?

There's no reason to have three separate files on disk, and certainly not three human-sync'd files.

> Programmatically modifying the yaml into three variants on disk before loading seems clunky and error prone to me.

Nothing requires multiple files on disk.  The main function should -- on the fly -- make whatever alterations to the _parsed scenario object_ that are required.  (It could still dump the revised object to disk for debugging, if desired).

**Comment by RussTedrake** [comment]
I understand.

Re: naming the ports. For input ports, I think that in the Apply* methods, it would be perfectly reasonable to call ExportInput with driver-specific names for any input port that is not already connected (e.g. because we opt-ed out of LCM).  And it would be perfectly reasonable to call ExportOutput with driver-specific names for all output ports.

Now that I finally understand your mental model, I can also see that we should have a single loop which calls Apply* everything (perhaps still MakeHardwareStation) and hardware_station_simulation should basically call MakeHardwareStation then simulator.AdvanceTo().
